File: examples/cpp/RecordReplay/record_video.cpp
#include <atomic>
#include <chrono>
#include <csignal>
#include <filesystem>
#include <iostream>
#include <memory>
#include <thread>

#include "depthai/depthai.hpp"

// Global flag for graceful shutdown
std::atomic<bool> quitEvent(false);

// Signal handler
void signalHandler(int signum) {
    quitEvent = true;
}

int main(int argc, char** argv) {
    // Parse command line arguments
    std::string outputFile = "test_video";
    for(int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if(arg == "-o" || arg == "--output") {
            if(i + 1 < argc) {
                outputFile = argv[++i];
            }
        }
    }

    // Set up signal handlers
    signal(SIGTERM, signalHandler);
    signal(SIGINT, signalHandler);

    try {
        // Create pipeline
        dai::Pipeline pipeline;

        // Create color camera node
        auto cam = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_A);

        // Create video encoder node
        auto videoEncoder = pipeline.create<dai::node::VideoEncoder>();
        videoEncoder->setProfile(dai::VideoEncoderProperties::Profile::H264_MAIN);
        cam->requestOutput(std::make_pair(1280, 720), dai::ImgFrame::Type::NV12)->link(videoEncoder->input);

        // Create record node
        auto record = pipeline.create<dai::node::RecordVideo>();
        record->setRecordVideoFile(std::filesystem::path(outputFile + ".avi"));
        videoEncoder->out.link(record->input);

        // Start pipeline
        pipeline.start();
        std::cout << "Recording video. Press Ctrl+C to stop." << std::endl;

        // Main loop
        while(pipeline.isRunning() && !quitEvent) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }

        // Cleanup
        pipeline.stop();
        pipeline.wait();

    } catch(const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: examples/python/RecordReplay/record_video.py
import depthai as dai
import argparse
import time
import signal
from pathlib import Path

parser = argparse.ArgumentParser()
parser.add_argument("-o", "--output", default="test_video", help="Output file name (without extension)")

args = parser.parse_args()

with dai.Pipeline() as pipeline:
    def signal_handler(sig, frame):
        print("Interrupted, stopping the pipeline")
        pipeline.stop()
    signal.signal(signal.SIGINT, signal_handler)

    cam = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_A)

    videoEncoder = pipeline.create(dai.node.VideoEncoder).build(cam.requestOutput((1280, 720), dai.ImgFrame.Type.NV12))
    videoEncoder.setProfile(dai.VideoEncoderProperties.Profile.H264_MAIN)

    record = pipeline.create(dai.node.RecordVideo)
    record.setRecordVideoFile(Path(f"{args.output}.mp4"))

    videoEncoder.out.link(record.input)

    pipeline.start()
    print("Recording video. Press Ctrl+C to stop.")
    while pipeline.isRunning():
        time.sleep(1)

